### ISSUE TITLE
Stepper motor library homing/limit capability

### ISSUE THREAD

**Issue by FineLineAutomation** [firmata/arduino#124]
I have been looking over the stepper motor library and was wondering if there was a way to stop a motor movement based on on a digital input going low. I have 2 use cases:

1) Using a stepper for position control you get to the end of physical motion and would like to put a limit switch in place that can stop the stepper.
2) Performing a homing operation on a stepper controlled motion axis with 

I could always perform #2 by bit banging everything from the PC, but would prefer to use a higher level command if necessary. 

I also thought about just doing it in hardware and just having the limit switch trip the enable pin on the Big Easy driver I am using, but you would need some way to manually control re-enable that  pin to get it to move off the switch.

**Comment by DomAmato** [comment]
I was playing with this for an implementation I needed and found a few things that seemed to work. The way I implemented it prevents the stepper motor from moving in the same direction but allows it to move in the other direction. It doesn't stop the motor though I probably should add that. Certainly the feature seems easy enough to add since the run state already includes a STOP which a command could just automatically set the runstate to effectively stopping the motor.

https://github.com/DomAmato/Robust-Firmata

Actually this could perhaps help me with a problem I ran into, if the pinModes weren't set in the constructor they didn't seem to work so I had to include it in the constructor. Is this caused by the implementation of the steppers being a pointer? I also had an issue with the encoder library not working as a pointer at all and had to add an empty constructor as a work around.

**Comment by soundanalogous** [comment]
Are you specifically talking about setting the pin modes for the limit pins? I see in RobustFirmata you're setting them to STEPPER, but they should be set to digital input or output since that is their function.

I'm also going to change the Stepper implementation at some point to use static rather than dynamic memory allocation.

Also if you want to propose an addendum to the stepper protocol, please open a pull request against: https://github.com/firmata/protocol/blob/master/stepper.md. See how I have done something for the I2C protocol: https://github.com/firmata/protocol/blob/master/i2c.md.

**Comment by DomAmato** [comment]
I am. In an earlier version I had a separate function that would add limit pins but within that function when I set pinMode to input_pullup it didn't enable the pin and would never work. Perhaps it was an error on my end but the code was pretty simple and for the most part is what I ended up moving over into the constructor. 

I am still tinkering with the functionality of the implementation so I have waited before proposing an addendum. I wanted to make sure it worked properly before I did that so that I didn't propose something and then change my own protocol.
